File: src/main/java/com/scalag/vulkan/memory/DescriptorSet.java
package com.scalag.vulkan.memory;

import com.scalag.vulkan.core.Device;
import com.scalag.vulkan.utility.VulkanAssertionError;
import com.scalag.vulkan.utility.VulkanObjectHandle;
import com.scalag.vulkan.compute.ComputePipeline;
import com.scalag.vulkan.core.Device;
import com.scalag.vulkan.utility.VulkanAssertionError;
import com.scalag.vulkan.utility.VulkanObjectHandle;
import org.lwjgl.system.MemoryStack;
import org.lwjgl.vulkan.VkDescriptorSetAllocateInfo;

import java.nio.LongBuffer;

import static org.lwjgl.system.MemoryStack.stackPush;
import static org.lwjgl.vulkan.VK10.*;

/**
 * @author MarconZet
 * Created 15.04.2020
 */
public class DescriptorSet extends VulkanObjectHandle {
    private final Device device;
    private final long descriptorSetLayout;
    private final DescriptorPool descriptorPool;

    public DescriptorSet(Device device, long descriptorSetLayout, DescriptorPool descriptorPool) {
        this.device = device;
        this.descriptorSetLayout = descriptorSetLayout;
        this.descriptorPool = descriptorPool;
        create();
    }

    @Override
    protected void init() {
        try (MemoryStack stack = stackPush()) {
            LongBuffer pSetLayout = stack.callocLong(1).put(0, descriptorSetLayout);
            VkDescriptorSetAllocateInfo descriptorSetAllocateInfo = VkDescriptorSetAllocateInfo.callocStack()
                    .sType(VK_STRUCTURE_TYPE_DESCRIPTOR_SET_ALLOCATE_INFO)
                    .descriptorPool(descriptorPool.get())
                    .pSetLayouts(pSetLayout);

            LongBuffer pDescriptorSet = stack.callocLong(1);
            int err = vkAllocateDescriptorSets(device.get(), descriptorSetAllocateInfo, pDescriptorSet);
            if (err != VK_SUCCESS) {
                throw new VulkanAssertionError("Failed to allocate descriptor set", err);
            }
            handle = pDescriptorSet.get();
        }
    }

    @Override
    protected void close() {
        vkFreeDescriptorSets(device.get(), descriptorPool.get(), handle);
    }
}

File: src/main/scala/com/scalag/vulkan/memory/DescriptorSet.scala
package com.scalag.vulkan.memory

import com.scalag.vulkan.compute.{Binding, LayoutSet}
import com.scalag.vulkan.core.Device
import com.scalag.vulkan.util.Util.{check, pushStack}
import com.scalag.vulkan.util.VulkanObjectHandle
import org.lwjgl.system.MemoryStack
import org.lwjgl.vulkan.VK10.*
import org.lwjgl.vulkan.{VkDescriptorBufferInfo, VkDescriptorSetAllocateInfo, VkWriteDescriptorSet}

/** @author
  *   MarconZet Created 15.04.2020
  */
class DescriptorSet(device: Device, descriptorSetLayout: Long, val bindings: Seq[Binding], descriptorPool: DescriptorPool) extends VulkanObjectHandle {

  protected val handle: Long = pushStack { stack =>
    val pSetLayout = stack.callocLong(1).put(0, descriptorSetLayout)
    val descriptorSetAllocateInfo = VkDescriptorSetAllocateInfo
      .calloc(stack)
      .sType$Default()
      .descriptorPool(descriptorPool.get)
      .pSetLayouts(pSetLayout)

    val pDescriptorSet = stack.callocLong(1)
    check(vkAllocateDescriptorSets(device.get, descriptorSetAllocateInfo, pDescriptorSet), "Failed to allocate descriptor set")
    pDescriptorSet.get()
  }
  
  def update(buffers: Seq[Buffer]): Unit = pushStack { stack =>
    val writeDescriptorSet = VkWriteDescriptorSet.calloc(buffers.length, stack)
    buffers.indices foreach { i =>
      val descriptorBufferInfo = VkDescriptorBufferInfo
        .calloc(1, stack)
        .buffer(buffers(i).get)
        .offset(0)
        .range(VK_WHOLE_SIZE)
      val descriptorType = buffers(i).usage match
        case storage if (storage & VK_BUFFER_USAGE_STORAGE_BUFFER_BIT) != 0 => VK_DESCRIPTOR_TYPE_STORAGE_BUFFER
        case uniform if (uniform & VK_BUFFER_USAGE_UNIFORM_BUFFER_BIT) != 0 => VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER
      writeDescriptorSet
        .get(i)
        .sType$Default()
        .dstSet(handle)
        .dstBinding(i)
        .descriptorCount(1)
        .descriptorType(descriptorType)
        .pBufferInfo(descriptorBufferInfo)
    }
    vkUpdateDescriptorSets(device.get, writeDescriptorSet, null)
  }

  override protected def close(): Unit =
    vkFreeDescriptorSets(device.get, descriptorPool.get, handle)
}
